glib/src/convert.rs

```diff
@@ -0,0 +1,292 @@
+            .unwrap_or_else(ptr::null_mut);
+        let conversions = unsafe {
+            ffi::g_iconv(
+                self.0,
+                &mut inbuf,
+                &mut inbytes_left,
+                &mut outbuf,
+                &mut outbytes_left,
+            )
+        };
+        if conversions as isize == -1 {
+            let err = io::Error::last_os_error();
+            let code = err.raw_os_error().unwrap();
+            if code == libc::EILSEQ || code == libc::EINVAL {
+                Err(IConvError::WithOffset {
+                    source: err,
+                    offset: input_len - inbytes_left,
+                })
+            } else {
+                Err(err.into())
+            }
+        } else {
+            Ok((conversions, inbytes_left, outbytes_left))
+        }
+    }
+}
+
+impl Drop for IConv {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::g_iconv_close(self.0);
+        }
+    }
+}
+
+#[doc(alias = "g_get_filename_charsets")]
+#[doc(alias = "get_filename_charsets")]
+pub fn filename_charsets() -> (bool, Vec<GString>) {
+    let mut filename_charsets = ptr::null_mut();
+    unsafe {
+        let is_utf8 = ffi::g_get_filename_charsets(&mut filename_charsets);
+        (
+            from_glib(is_utf8),
+            FromGlibPtrContainer::from_glib_none(filename_charsets),
+        )
+    }
+}
+
+#[doc(alias = "g_locale_from_utf8")]
+pub fn locale_from_utf8(utf8string: &GStr) -> Result<(Slice<u8>, usize), CvtError> {
+    assert!(utf8string.len() <= isize::MAX as usize);
+    let mut bytes_read = 0;
+    let mut bytes_written = 0;
+    let mut error = ptr::null_mut();
+    let result = unsafe {
+        ffi::g_locale_from_utf8(
+            utf8string.as_ptr(),
+            utf8string.len() as isize,
+            &mut bytes_read,
+            &mut bytes_written,
+            &mut error,
+        )
+    };
+    if result.is_null() {
+        Err(CvtError::new(unsafe { from_glib_full(error) }, bytes_read))
+    } else {
+        let slice = unsafe { Slice::from_glib_full_num(result, bytes_written as usize) };
+        Ok((slice, bytes_read))
+    }
+}
+
+#[doc(alias = "g_filename_to_utf8")]
+pub fn filename_to_utf8(
+    opsysstring: impl AsRef<std::path::Path>,
+) -> Result<(crate::GString, usize, usize), crate::Error> {
+    let path = opsysstring.as_ref().to_glib_none();
+    unsafe {
+        let mut bytes_read = std::mem::MaybeUninit::uninit();
+        let mut bytes_written = std::mem::MaybeUninit::uninit();
+        let mut error = ptr::null_mut();
+        let ret = ffi::g_filename_to_utf8(
+            path.0,
+            path.1.as_bytes().len() as isize,
+            bytes_read.as_mut_ptr(),
+            bytes_written.as_mut_ptr(),
+            &mut error,
+        );
+        if error.is_null() {
+            Ok((
+                from_glib_full(ret),
+                bytes_read.assume_init(),
+                bytes_written.assume_init(),
```

Comment:
Isn't this equivalent to the length of the returned `GString`? And I assume it can be used also to give a size hint to it when creating the `GString`

Reply:
Only bytes_read works as a size hint, the bytes_written param does not have to be returned though.